### PR TITLE
calculating PartSize based on size of upload

### PR DESCRIPTION
**Pull request by trickkiste**
This is a fix for bug #149.

It has been applied to latest commit in koenvo:s3-large-file-upload-fix branch, not to master and therefore shall be merged after the former has been merged successfully.

So this pull request might be premature, but I wanted to give you a chance to review it early and hope to get some feedback, to get it merged quickly.

My best regards,
Markus

### CODE REVIEW

File: s3store/s3store.go [hunk 1]
```diff
@@ -536,3 +549,21 @@ func isAwsError(err error, code string) bool {
 	}
 	return false
 }
+
+func (store S3Store) CalcPartSize(size *int64) (effectivePartSize int64) {
```
Comment (Acconut) [inline s3store/s3store.go]: Any reason why you decided to use a pointer here?

File: s3store/s3store.go [hunk 2]
```diff
@@ -259,11 +256,11 @@ func (store S3Store) WriteChunk(id string, offset int64, src io.Reader) (int64,
 			return bytesUploaded, nil
 		}
 
-		if (size - offset) <= store.MinPartSize {
+		if (size - offset) <= effectivePartSize {
```
Comment (Acconut) [inline s3store/s3store.go]: Are you sure that **Min**PartSize has been replaced here correctly?

File: s3store/s3store.go [hunk 3]
```diff
@@ -259,11 +256,11 @@ func (store S3Store) WriteChunk(id string, offset int64, src io.Reader) (int64,
...
 			if (size - offset) != n {
 				return bytesUploaded, nil
 			}
-		} else if n < store.MinPartSize {
+		} else if n < effectivePartSize {
```
Comment (Acconut) [inline s3store/s3store.go]: Are you sure that **Min**PartSize has been replaced here correctly?

File: s3store/s3store.go [hunk 4]
```diff
@@ -536,3 +549,21 @@ func isAwsError(err error, code string) bool {
...
+	case *size/10000 < store.MinPartSize:
+		effectivePartSize = store.MinPartSize
+	case int64(math.Mod(float64(*size), 10000)) == 0:
+		effectivePartSize = *size / 10000
+	case *size/9999 >= store.MaxPartSize:
+		effectivePartSize = store.MaxPartSize
+	default:
+		effectivePartSize = *size / 9999
+	}
+	return effectivePartSize
```
Comment (Acconut) [inline s3store/s3store.go]: I am sorry but I am lost with this switch case here. Could you explain why you ditched your proposed algorithm from https://github.com/tus/tusd/issues/149?

File: s3store/s3store.go (hunk 2 above)
Comment (trickkiste) [inline s3store/s3store.go]: With this new approach, we are calculating an optimal effectivePartSize, which needs to respect the MinPartSize and MaxPartSize limits set in the S3 API specification and take into account absolute max-parts limit of 10.000 parts. However, after the point of calculation, the min and max values are not relevant for our transfer anymore.

Each part we upload will need to align to effectivePartSize except for the last one, which can and will in most cases be lower than effectivePartSize and can also be lower than MinPartSize. Still, whether it is actually lower than MinPartSize does not have any relevance anymore.

However, the parts uploaded prior to the last one, must never be smaller than effectivePartSize, as we either would have to increase the size of another part to compensate for that, or we would need to upload an additional part in order to get everything transferred within the absolute max-parts limit of 10.000 parts. Otherwise we would have to exceed the absolute parts-max limit of 10.000 parts by at least one.

So after the point of calculation of the optimal part size (to suffice all restrictions in S3 API), MinPartSize and MaxPartSize loose all their relevance, as effectivePartSize does respect these limits anyways.

File: s3store/s3store.go (hunk 3 above)
Comment (trickkiste) [inline s3store/s3store.go]: Same as above.

File: s3store/s3store.go [hunk 5]
```diff
@@ -148,7 +149,7 @@ func New(bucket string, service S3API) S3Store {
 	return S3Store{
 		Bucket:      bucket,
 		Service:     service,
-		MaxPartSize: 6 * 1024 * 1024,
+		MaxPartSize: 5 * 1024 * 1024 * 1024,
```
Comment (koenvo) [inline s3store/s3store.go]: This should be `525 * 1024 * 1024`. (5TB / 10.000 parts)

File: s3store/s3store.go (hunk 4 above)
Comment (trickkiste) [inline s3store/s3store.go]: The proposed "algorithm" was rather a rough human language description of what needs to be implemented, than an actual algorithm. The switch statement here does implement that outline and I have tested it thoroughly. The test case does test all relevant corner cases. But I have run the code for all sizes from MaxPartSize*9999-1 upwards.

File: s3store/s3store.go (hunk 1 above)
Comment (trickkiste) [inline s3store/s3store.go]: Efficiency. Mainly for my "testing the code for ALL possible file sizes from 1 Byte to 5 TB" experiment.

File: s3store/s3store.go (hunk 5 above)
Comment (trickkiste) [inline s3store/s3store.go]: @koenvo good point!

While I would propose to solve that differently, as going your way causes confusion, since the S3 API spec states a MaxPartSize of "5 * 1024 * 1024 * 1024" and new contributors will not intuitively understand, why that is different here. 

I would therefore propose to add two other variables here and adapt calcPartSize take them into account, which - as you correctly found out - is not the case now as far as MaxUploadSize goes.

```
MaxUploadParts: 10000
 MaxUploadSize: 5 * 1024 * 1024 * 1024 * 1024
```

Comment (trickkiste) [inline s3store/s3store.go]: I would also adapt the tests to test for this limit, so mistakes like that one can not happen anymore.

File: s3store/s3store.go (hunk 4 above)
Comment (trickkiste) [inline s3store/s3store.go]: I will add some documentation of what is actually going on there.

File: s3store/s3store.go (hunk 1 above)
Comment (trickkiste) [inline s3store/s3store.go]: If you don't like it, we can drop it.
If there is a good reason to refrain from it, then please elaborate, as I am always eager to learn something new!

File: s3store/s3store.go (hunk 4 above)
Comment (trickkiste) [inline s3store/s3store.go]: I have improved on the algorithm, simplified it as much as possible and made sure it will still work, even if Amazon changes current limits in S3 API spec (or if other implementations might use different values).

I have adapted existing tests and improved my own testcase. `go fmt ./s3store` has been run on the code. The patch is also based on the current master which already includes koenvo's patch.

Please let me know, if you need anything else done on this. I am eager to get this merged.

**Comment by trickkiste** [comment]
@Acconut I have taken all comments into consideration and came up with an improved implementation. The fact, that in case of interruption the last part needs to be re-transmitted, remains. But that was the case before already.

However, I will keep thinking about this problem and there might be a way to fix this, somewhere down the direction you proposed earlier. But it is going to be a mayor rewrite and until then, I suppose it is best to go with the current patch, so that we have a system, able to successfully handle files up to MaxObjectSize, until we come up with a better way to do the same.

File: s3store/s3store.go [hunk 6]
```diff
@@ -547,3 +561,28 @@ func isAwsError(err error, code string) bool {
 	}
 	return false
 }
+
+func (store S3Store) CalcOptimalPartSize(size *int64) int64 {
```
Comment (Acconut) [inline s3store/s3store.go]: Any reason why you are using a pointer as the argument here instead of just the scalar `int64`?

File: s3store/s3store.go [hunk 7]
```diff
@@ -547,3 +561,28 @@ func isAwsError(err error, code string) bool {
...
+	switch {
+	// We can only manage files up to MaxObjectSize, else we need to fail.
+	case *size > store.MaxObjectSize:
+		return 0
```
Comment (Acconut) [inline s3store/s3store.go]: In such a case, an explicit error should be returned.

File: s3store/s3store.go [hunk 8]
```diff
@@ -547,3 +561,28 @@ func isAwsError(err error, code string) bool {
...
+	// When we need 9999 parts or less with MinPartSize.
+	case *size/store.MinPartSize < store.MaxMultipartParts:
+		return store.MinPartSize
+	// When we can divide our upload into exactly MaxMultipartParts parts with
+	// no bytes leftover, we will not need an spare last part.
+	// Also, when MaxObjectSize is equal to MaxPartSize * MaxMultipartParts
+	// (which is not the case with the current AWS S3 API specification, but
+	// might be in the future or with other S3-aware stores), we need this in
+	// order for our Multipart-Upload to reach full MaxObjectSize.
+	case int64(math.Mod(float64(*size), float64(store.MaxMultipartParts))) == 0:
```
Comment (Acconut) [inline s3store/s3store.go]: I believe this case could be merged with the default one since it should not make a difference here whether we need a spare last part or not. But maybe I am missing something.

Comment (trickkiste) [inline s3store/s3store.go]: If we go without this, we will not be able to fully upload a file with MaxObjectSize, in case MaxPartSize*MaxMultipartParts == MaxObjectSize. I would keep this to be prepared for eventual changes to the current S3 specs, which has happened on multiple occasions in the past, as far as I remember. That way, we are prepared for all eventualities.

File: s3store/s3store.go (hunk 6 above)
Comment (trickkiste) [inline s3store/s3store.go]: I was writing a test, which would run this function for ALL possible file sizes. Therefore I was trying to be as efficient on memory-use as possible, as these are quite a lot of test cases.

Comment (trickkiste) [inline s3store/s3store.go]: Do you want me to change that?

File: s3store/s3store.go (hunk 7 above)
Comment (trickkiste) [inline s3store/s3store.go]: OK, I was a little lazy there.

File: s3store/s3store.go (hunk 6 above)
Comment (Acconut) [inline s3store/s3store.go]: > Therefore I was trying to be as efficient on memory-use as possible, as these are quite a lot of test cases.

Please don't do this kind of premature optimization. Unless you have proven using a benchmark that this change provides an improvement, stay away for that. Code readability and ease to understanding it is usually more important than performance.

Also I have serious doubt that this would bring any speed-ups since you need to dereference the pointer multiple times and also need to allocate memory for the pointer which could mean even worse memory-efficient. However, I am not a 100% sure on how Golang handles stack pointers but it doesn't matter here.

> Do you want me to change that?

Yes, please do so.

File: s3store/s3store.go [hunk 9]
```diff
@@ -547,3 +561,28 @@ func isAwsError(err error, code string) bool {
...
+		return *size / store.MaxMultipartParts
+	// In all other cases, we need a spare last piece for the remaining bytes.
+	default:
+		return *size / (store.MaxMultipartParts - 1)
```
Comment (Acconut) [inline s3store/s3store.go]: Why is the -1 here? The last spare part can also have a size of up to `size / MaxMultipartParts` but it just won't all be used. Also we need to double check whether this integer division is not rounding down sometimes or else uploads would not finish.

Comment (trickkiste) [inline s3store/s3store.go]: I am afraid, you misread `MaxMultipartParts` here. The "- 1" in `(store.MaxMultipartParts - 1) `does not reduce the part size, but the total number of parts, we are dividing up our upload by. The part size actually grows due to that.

- MaxMultipartParts
is the maximum number of parts an S3 multipart upload is allowed to have according to AWS S3 API specifications.

However, compiling this answer, I actually found a bug in the algorithm, another special case, which needs to be checked for. I'll attend to that at the bottom of this post.

As for math.Mod, I just realised, there is the "%" operator for the remainder of an integer division, so of course that one is better suited, as it produces predictable results with no rounding problems. I will switch over to that.

As for the algorithm per se, the division by `MaxMultipartParts` will almost never produce an exact partition, but a number rounded down to an integer and there will be a remainder to this division. That remainder is what will end up in the last spare part. As we expect there to be a remainder, we need to divide the upload size by 9999 not 10.000, so we have a part left for the remainder. That is what is going on here.

The section above the `default:` checks, whether a division by 10.000 is applicable, by checking if a division `size / MaxMultipartParts` will produce any remainder or not. And only if it is a division without a remainder, we divide by `MaxMultipartParts` not `(MaxMultipartParts-1)` (10.000 not 9.999 in our case).

Here are a few examples with simple numbers for illustration:
MaxMultipartParts = 10
MinPartSize = 5
MaxPartSize = 10
MaxObjectSize = 100

size = 100
size / MaxMultipartParts  
= 100 / 10 = 10 with no remainder 
... No remainder, so we will divide by MaxMultipartParts and have 10 parts with size 10.
Total: 10 parts - SUCCESS.

size = 98
size / MaxMultipartParts  
= 99 / 10 = 9 with a remainder of 8
... So if we would use the result of this division as our optimalPartSize, then we would end up with 10 parts of size 9 and an 11th part of size 8. Which does not work, as we would exceed MaxMultipartParts.

You could say, well then let's just use the integer result of this division as the optimalPartSize (ergo 9). The fact that this division seemingly already produces, what we expect as optimalPartSize, is only a result of the fact, that MaxMultipartParts and MaxPartSize both are 10 in our example. So don't be mislead and rather ignore that fact.

Next we divide the size by (MaxMultipartParts - 1), so we have one spare part left, just in case.
size = 98
size / (MaxMultipartParts - 1)
= 99 / 9 = 10 with a remainder of 8
... So we have 9 parts of size 10 and a 10th part of size 8. 
Total: 10 parts - SUCCESS

I skipped size = 99, as it is a special case, and actually reveals a bug in the current algorithm:
size = 99
size / MaxMultipartParts
= 99 / 10 = 9 with a remainder of 9
... That means, we would end up with 10 parts of size 9 and an 11th part of size 9 again. In the current algorithm, as we find a rest in this division, we need to progress to the next case. 

But if we divide by (MaxMultipartParts - 1), we wont be lucky as well in this case:
= 99 / 9 = 11 with no remainder in that special case
... we end up with 9 parts of size 11, which exceeds MaxPartSize.

So we need to add another check:
case: size % (MaxMultiparts - 1) > MaxPartSize 
(ergo, if the last part would be larger than MaxPartSize)
then optimalPartSize = MaxPartSize
... we end up with 9 parts of size 10 and a last part of size 9. 
Total: 10 parts - SUCCESS

I'll play that with a few other numbers:
MaxMultipartParts = 20
MinPartSize = 5
MaxPartSize = 10
MaxObjectSize = 200

size 200
200 / 20 = 10 no remainder
... 20 parts with size 10. 
Total: 20 parts - SUCCESS.

size 199
199 / 20 = 9 with a remainder of 19
199 / 19 = 10 with a remainder of 9. The remainder is less than MaxPartSize, so apply the default.
... 19 parts with size 10 and a last one with size 9. 
Total: 20 parts - SUCCESS.

size 180
180 / 20 = 9 with no remainder
... 20 parts with size 9. 
Total: 20 parts - SUCCESS.

size 181
181 / 20 = 9 with a remainder of 1
181 / 19 = 9 with a remainder of 10. The remainder is not greater than MaxPartSize, so we apply default.
... 19 parts with size 9 and a last one with size 10. 
Total: 20 parts - SUCCESS.

We could also apply optimalPartSize = MaxPartSize here, so we would end up with 18 parts with size 10 and a last one with size 1. Total: 19 parts. As we thrive for the maximum number of parts with a minimum PartSize, the former solution is better.

size 182
182 / 20 = 9 with a remainder of 2
182 / 19 = 9 with a remainder of 11. The remainder is greater than MaxPartSize, so we apply MaxPartSize.
... 18 parts with size 10 and a last one with size 2. 
Total: 19 parts - SUCCESS.

size 189
189 / 20 = 9 with a remainder of 9
189 / 19 = 9 with a remainder of 18. The remainder is greater than MaxPartSize, so we apply MaxPartSize.
... 18 parts with size 10 and a last one with size 9. 
Total: 19 parts - SUCCESS.

size 190
190 / 20 = 9 with a remainder of 10 remainder
190 / 19 = 10 with no remainder. The remainder is less than MaxPartSize, so we apply default.
... 19 parts with size 10. 
Total: 19 parts - SUCCESS.

size 191
191 / 20 = 9 with a remainder of 11. We have a remainder, so we proceed to next case.
191 / 19 = 10 with a remainder of 1. The remainder is less than MaxPartSize, so we apply default.
... 19 parts with size 10 and a last one with size 1. 
Total: 20 parts - SUCCESS.

The addition of this additional switch should produce a universally applicable algorithm and replacing math.Mod with "%" operator, will fix your other concern.

Comment (trickkiste) [inline s3store/s3store.go]: I also need to test for lastpart > MaxPartSize in my tests.

**Comment by trickkiste** [comment]
@Acconut, I cleaned up and simplified the algorithm.

Added error handling also. That part might need revision, as I don't know yet, what good standards are for doing that.

And I added a few more test patterns.

I tested the algorithm with a bunch of different settings for Min/MaxPartSize, MaxObjectSize and MaxMultipartParts. All produced positive results.

File: s3store/s3store.go [hunk 10]
```diff
@@ -547,3 +560,40 @@ func isAwsError(err error, code string) bool {
...
+
+func (store S3Store) CalcOptimalPartSize(size int64) (optimalPartSize int64, err error) {
+	var errorstrings []string
+
+	switch {
+	// When upload is smaller or equal MinPartSize, we upload in just one part.
+	case size <= store.MinPartSize:
+		optimalPartSize = store.MinPartSize
+	// When we need 9999 parts or less with MinPartSize.
+	case size/store.MinPartSize < store.MaxMultipartParts:
```
Comment (Acconut) [inline s3store/s3store.go]: Could we remove the integer division here and test for `size < MinPartSize*MaxMultipartParts` instead? I don't think this makes a difference here but I would like to avoid integer division whenever possible since it does not behave intuitively first :)

File: s3store/s3store.go [hunk 11]
```diff
@@ -547,3 +560,40 @@ func isAwsError(err error, code string) bool {
...
+		optimalPartSize = store.MinPartSize
+	// If our upload divides up exactly into MaxMultipartParts parts with
+	// no bytes leftover, we will not need an spare last part. So we can go with
+	// a straight division. Otherwise we would exceed MaxPartSize in a scenario,
+	// where MaxObjectSize is equal to MaxPartSize * MaxMultipartParts
+	// (which is not the case with the current AWS S3 API specification, but
+	// might be in the future or with other S3-aware stores).
+	case size%store.MaxMultipartParts == 0:
+		optimalPartSize = size / store.MaxMultipartParts
+	// In all other cases, we need to round up to the next integer.
```
Comment (Acconut) [inline s3store/s3store.go]: Would you be so kind and add to the comment that integer division will round down, if necessary. This will help understanding the "+1"

Comment (trickkiste) [reply s3store/s3store.go]: No problem.

File: s3store/s3store.go [hunk 12]
```diff
@@ -562,27 +561,39 @@ func isAwsError(err error, code string) bool {
...
 	default:
-		return *size / (store.MaxMultipartParts - 1)
+		optimalPartSize = size/store.MaxMultipartParts + 1
+	}
+
+	// an upload larger than MaxObjectSize must throw an error
+	if size > store.MaxObjectSize {
+		errorstrings = append(errorstrings, fmt.Sprintf("CalcOptimalPartSize: upload size of %v bytes exceeds MaxObjectSize of %v bytes", size, store.MaxObjectSize))
```
[the remark this answers is not in the document]
Comment (trickkiste) [reply s3store/s3store.go]: I wanted optimalPartSize still to be calculated and returned in any case, as it helps to debug, how the algorithm is behaving. Especially useful for the tests.

File: s3store/s3store.go (hunk 10 above)
Comment (trickkiste) [reply s3store/s3store.go]: We can do that. My emphasis lies on readability and understandability and this change does rather further them then harm them.

**Comment by trickkiste** [comment]
@Acconut, I am all for running the algorithm for all file sizes from `0 to MaxObjectSize+1` with the settings you proposed.
```
MaxPartSize:       5 * 1024,
MinPartSize:       5,
MaxMultipartParts: 10000,
MaxObjectSize:     5 * 1024 * 1024,
```
Except, I would probably go for 

> MaxMultipartParts: 1000

 or at least 

> MaxObjectSize: MaxPartSize*MaxMultipartParts

 to cover that edge case as well.

I had run such a simulation with a prior version of the algorithm and less test cases. So I would have to run it anew, as we have had a couple of iterations since. I would be equally happy, if you would run it ;-)

File: s3store/s3store.go (hunk 11 above)
Comment (trickkiste) [reply s3store/s3store.go]: Doing the division in float64 and then rounding up the result is no option I guess, as we might end up with rounding errors due to the float nature of things?!

It would enhance readability, but probably introduce a whole new set of problems.

Comment (Acconut) [reply s3store/s3store.go]: I believe the errors which occur due to the use of a floating point number is too minimal to be harmful for our usecase but I am not keen on testing that :)
Also, rounding functions in Go also returns a float64 not integers. Let's keep the current approach.

File: s3store/s3store.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (Acconut) [reply s3store/s3store.go]: IMO there is no need to calculate a part size if the upload exceeds the maximum object size because the upload will never be possible. It's confusion why this check is down here.

[the remark this answers is not in the document]
Comment (trickkiste) [reply s3store/s3store.go]: OK. So I return part size 0 and the error then. Coming in a second.

[the remark this answers is not in the document]
Comment (trickkiste) [reply s3store/s3store.go]: I will have to restructure the tests again to accommodate that change. One more commit coming.

[the remark this answers is not in the document]
Comment (trickkiste) [reply s3store/s3store.go]: I am too quick to push. One more fix pending.

File: s3store/s3store_test.go [hunk 13]
```diff
@@ -24,6 +24,138 @@ var _ tusd.TerminaterDataStore = s3store.S3Store{}
...
+			err += fmt.Sprintf("Testcase #%v size %v: size exceeds MaxObjectSize %v but no error returned\n", index, size, MaxObjectSize)
+		}
+		if debug && optimalPartSize == 0 {
+			fmt.Printf("Testcase #%v size %v: size exceeds MaxObjectSize %v\n", index, size, MaxObjectSize)
+		}
+		if optimalPartSize > 0 {
+			equalparts = size / optimalPartSize
+			lastpartsize = size % optimalPartSize
+			if optimalPartSize < MinPartSize {
+				err += fmt.Sprintf("Testcase #%v size %v, %v parts of size %v, lastpart %v: optimalPartSize < MinPartSize %v\n", index, size, equalparts, optimalPartSize, lastpartsize, MinPartSize)
```
Comment (Acconut) [inline s3store/s3store_test.go]: Any reason why you are not using t.Errorf here and in the following if statement?

File: s3store/s3store_test.go [hunk 14]
```diff
@@ -24,6 +24,138 @@ var _ tusd.TerminaterDataStore = s3store.S3Store{}
...
+	var HighestApplicablePartSize int64 = MaxObjectSize / MaxMultipartParts
+	if MaxObjectSize%MaxMultipartParts > 0 {
+		HighestApplicablePartSize++
+	}
+	var RemainderWithHighestApplicablePartSize int64 = MaxObjectSize % HighestApplicablePartSize
+
+	// some of these tests are actually duplicates, as they specify the same size
+	// in bytes - two ways to describe the same thing. That is wanted, in order
+	// to provide a full picture from any angle.
+	testcases := []int64{
```
Comment (Acconut) [inline s3store/s3store_test.go]: It would be nice if the test cases where separated into situations where a part size can and cannot be calculated, making the result assertion and error checking for straight forward.

File: s3store/s3store_test.go (hunk 13 above)
Comment (trickkiste) [reply s3store/s3store_test.go]: No. As said, with error handling, I am hoping for your advice. My Golang experience is limited and I did not read up on everything yet. I was generating strings and only using t.Errorf at the bottom, to make sure, I see all errors not just the first one. Don't know, if this is necessary.

File: s3store/s3store_test.go (hunk 14 above)
Comment (trickkiste) [reply s3store/s3store_test.go]: It is true, that the current set of checks in TestCalcOptimalPartSize are not straight forward.

However, in my opinion the assertions should work with any size you throw at them, no matter, if we expect a success of an error from CalcOptimalPartSize. One reason for that is, that certain testcases might produce the one result with set A of S3 settings but might produce another with set B.

Before we moved the MaxObjectSize condition above the calculation of the optimalPartSize and by that means creating a scenario, where CalcOptimalPartSize could output 0 as a result, the tests were actually very straight forward, because they were linear. No if condition in there. That is the reason, why I preferred optimalPartSize being calculated, even if size > MaxObjectSize.

One could also argue, that the purpose of CalcOptimalPartSize is calculating said part size and checking whether the size of the upload is larger than MaxObjectSize should not even concern this function but actually would belong somewhere else!

E.g. in NewUpload, because why would we create such an upload in the first place, if we could never finish it.

Therefore I would propose to move that condition to NewUpload and keep CalcOptimalPartSize and it's tests clean.

Don't you agree?

**Comment by trickkiste** [comment]
Would you share your current test setup? I'd like to collaborate.

**Comment by trickkiste** [comment]
I pushed my proposal on how to simplify CalcOptimalPartSize.

File: s3store/s3store_test.go (hunk 13 above)
Comment (Acconut) [reply s3store/s3store_test.go]: That's why I asked :) FYI, t.Error will continue execution after it has been invoked while t.Fatal will immediately stop the test case (see https://golang.org/pkg/testing/#B.Errorf). For example, I use t.Fatal if a database connection could not be made and further testing is useless. On the other hand, t.Error is used for asserting that the result has specific properties.

File: s3store/s3store_test.go [hunk 15]
```diff
@@ -43,119 +43,129 @@ func TestCalcOptimalPartSize(t *testing.T) {
...
+
+	var debug = false
+	var equalparts, lastpartsize int64
+
+	// sanity check
+	if store.MaxObjectSize > store.MaxPartSize*store.MaxMultipartParts {
+		t.Errorf("MaxObjectSize %v can never be achieved, as MaxMultipartParts %v and MaxPartSize %v only allow for an upload of %v bytes total.\n", store.MaxObjectSize, store.MaxMultipartParts, store.MaxPartSize, store.MaxMultipartParts*store.MaxPartSize)
+	}
+
+	for size := int64(1); size <= store.MaxObjectSize; size++ {
```
Comment (Acconut) [inline s3store/s3store_test.go]: Please start with `size = 0`. In theory, empty uploads are also allowed.

File: s3store/s3store_test.go [hunk 16]
```diff
@@ -43,119 +43,129 @@ func TestCalcOptimalPartSize(t *testing.T) {
...
+			fmt.Printf("Size %v, %v parts of size %v, lastpart %v, does exceed MaxObjectSize %v.\n", size, equalparts, optimalPartSize, lastpartsize, size > store.MaxObjectSize)
 		}
 	}
 	// fmt.Println("HighestApplicablePartSize", HighestApplicablePartSize)
 	// fmt.Println("RemainderWithHighestApplicablePartSize", RemainderWithHighestApplicablePartSize)
 }
 
+func TestAllPartSizes(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping test in short mode.")
```
Comment (Acconut) [inline s3store/s3store_test.go]: Skipping this test is not necessary I believe. For me it completes in 0.2s.

File: s3store/s3store_test.go [hunk 17]
```diff
@@ -43,119 +43,129 @@ func TestCalcOptimalPartSize(t *testing.T) {
...
+	// sanity check
+	if store.MaxObjectSize > store.MaxPartSize*store.MaxMultipartParts {
+		t.Errorf("MaxObjectSize %v can never be achieved, as MaxMultipartParts %v and MaxPartSize %v only allow for an upload of %v bytes total.\n", store.MaxObjectSize, store.MaxMultipartParts, store.MaxPartSize, store.MaxMultipartParts*store.MaxPartSize)
+	}
+
+	for size := int64(1); size <= store.MaxObjectSize; size++ {
+		optimalPartSize, calcError := store.CalcOptimalPartSize(size)
+		equalparts = size / optimalPartSize
+		lastpartsize = size % optimalPartSize
+		assert.False(optimalPartSize < store.MinPartSize, fmt.Sprintf("Size %v, %v parts of size %v, lastpart %v: optimalPartSize < MinPartSize %v.\n", size, equalparts, optimalPartSize, lastpartsize, store.MinPartSize))
```
Comment (Acconut) [inline s3store/s3store_test.go]: You can save the call to fmt.Sprintf by using assert.Falsef  and similar methods (see https://godoc.org/github.com/stretchr/testify/assert#Assertions.Falsef).

File: s3store/s3store_test.go [hunk 18]
```diff
@@ -43,119 +43,129 @@ func TestCalcOptimalPartSize(t *testing.T) {
...
+	var debug = false
+	var equalparts, lastpartsize int64
+
+	// sanity check
+	if store.MaxObjectSize > store.MaxPartSize*store.MaxMultipartParts {
+		t.Errorf("MaxObjectSize %v can never be achieved, as MaxMultipartParts %v and MaxPartSize %v only allow for an upload of %v bytes total.\n", store.MaxObjectSize, store.MaxMultipartParts, store.MaxPartSize, store.MaxMultipartParts*store.MaxPartSize)
+	}
+
+	for size := int64(1); size <= store.MaxObjectSize; size++ {
+		optimalPartSize, calcError := store.CalcOptimalPartSize(size)
```
Comment (Acconut) [inline s3store/s3store_test.go]: Is there a reason why you didn't check for `calcError == nil` here? For the range of `size` is should never return an error.

File: s3store/s3store_test.go (hunk 16 above)
Comment (trickkiste) [reply s3store/s3store_test.go]: Well for you.
31.883s for me!
MacBook Air Mid 2012

File: s3store/s3store_test.go (hunk 18 above)
Comment (trickkiste) [reply s3store/s3store_test.go]: Yes there is.
I view the range we are testing here as arbitrary and therefore the assertions should work with any range you choose. I actually would much prefer to test up to MaxObjectSize+1. An assertion `calcError == nil` would not work in this case.

File: s3store/s3store_test.go (hunk 15 above)
Comment (trickkiste) [reply s3store/s3store_test.go]: Roger that.

File: s3store/s3store_test.go (hunk 18 above)
Comment (trickkiste) [reply s3store/s3store_test.go]: Also I believe that having identical assertions in the two tests concerning CalcOptimalPartSize helps prevent confusion. As we are testing the same thing, just different ranges of test cases, so why should the test differ.

Ideally, we would externalise the assertions into a function, which we use in both tests, which would assure, they stay the same over time. But as this would complicate things instead of making them more readable, I did refrain from trying that.

File: s3store/s3store_test.go (hunk 17 above)
Comment (trickkiste) [reply s3store/s3store_test.go]: When I do this: 
```
assert.Falsef(optimalPartSize < store.MinPartSize, "Size %v, %v parts of size %v, lastpart %v: optimalPartSize < MinPartSize %v.\n", size, equalparts, optimalPartSize, lastpartsize, store.MinPartSize)
```

I get that: 
```
assert.Falsef undefined (type *assert.Assertions has no field or method Falsef)
```

With `go version go1.7.6 linux/amd64`

Comment (Acconut) [reply s3store/s3store_test.go]: That's because the vendored version of testify (https://github.com/tus/tusd/tree/master/vendor/github.com/stretchr/testify) is not a recent one and does not feature the Falsef method (alongside others). I will update the dependency tomorrow on master so you can merge it into this branch.

Comment (trickkiste) [reply s3store/s3store_test.go]: OK, I'll merge the new master then.
But let's get this to the finishing line soon now.
I am finishing the TestNewUploadLargerMaxObjectSize test and that should be it then.
Or are you missing anything else?

File: s3store/s3store_test.go (hunk 18 above)
Comment (Acconut) [reply s3store/s3store_test.go]: I disagree. Call it preference or not, I am not a fan of having different scenarios into one test.

>  I actually would much prefer to test up to MaxObjectSize+1

I don't see it this way. This should be its own separate test since the expected outcome and the used assertions are different. Furthermore, this will make debugging issues easier since you can run a single test isolated using `go test -run`.

File: s3store/s3store_test.go (hunk 14 above)
Comment (Acconut) [reply s3store/s3store_test.go]: > However, in my opinion the assertions should work with any size you throw at them, no matter, if we expect a success or an error from CalcOptimalPartSize. 

I don't share this opinion. Test should assert the returned output from a function if you provide it with the specific input, making them straightforward to understand and also debug if necessary. For example, let's take `store.MaxObjectSize + 1`. Should a part size be calculated for this input? I don't know and I cannot tell from this test case. Frankly, I cannot even tell you whether your test verifies that the behavior of CalcPartSize does not change.

>  One reason for that is, that certain testcases might produce the one result with set A of S3 settings but might produce another with set B. Therefore having dividing the test into a set, which is expected to fail, and one, which is expected to succeed, might not actually be realistic.

If you make the test inputs depend on store.MaxPartSize (etc.) then this should not be an issue, as you did.

> Therefore I would propose to move that condition to NewUpload and keep CalcOptimalPartSize and it's tests clean and universal.

Seen that you already took care of this, I welcome this change :)

File: s3store/s3store_test.go [hunk 19]
```diff
@@ -150,16 +151,16 @@ func TestAllPartSizes(t *testing.T) {
...
-		assert.False(size%optimalPartSize == 0 && equalparts > store.MaxMultipartParts, fmt.Sprintf("Size %v, %v parts of size %v, lastpart %v: more parts than MaxMultipartParts %v.\n", size, equalparts, optimalPartSize, lastpartsize, store.MaxMultipartParts))
-		assert.False(size%optimalPartSize > 0 && equalparts > store.MaxMultipartParts-1, fmt.Sprintf("Size %v, %v parts of size %v, lastpart %v: more parts than MaxMultipartParts %v.\n", size, equalparts, optimalPartSize, lastpartsize, store.MaxMultipartParts))
-		assert.False(lastpartsize > store.MaxPartSize, fmt.Sprintf("Size %v, %v parts of size %v, lastpart %v: lastpart > MaxPartSize %v.\n", size, equalparts, optimalPartSize, lastpartsize, store.MaxPartSize))
-		assert.False(lastpartsize > optimalPartSize, fmt.Sprintf("Size %v, %v parts of size %v, lastpart %v: lastpart > optimalPartSize %v.\n", size, equalparts, optimalPartSize, lastpartsize, optimalPartSize))
+		assert.False(optimalPartSize < store.MinPartSize, "Size %v, %v parts of size %v, lastpart %v: optimalPartSize < MinPartSize %v.\n", size, equalparts, optimalPartSize, lastpartsize, store.MinPartSize)
```
Comment (Acconut) [inline s3store/s3store_test.go]: They have to be assert.Falsef calls, I believe.

File: s3store/s3store_test.go (hunk 18 above)
Comment (trickkiste) [reply s3store/s3store_test.go]: Well, that is certainly a valid argument.

But if one would follow that perspective, the test would need to be ripped apart even further. Each one of the case statements in CalcOptimalPartSize then would represent its own test case in your scenario. 

Readability will be totally gone then. 

No single point to understand, what we are expecting from CalcOptimalPartSize anymore. And you would have to define universal scenarios for sizes, which will trigger the expected behaviour. Which will become rather complicated, when you consider, that MinPartSize, MaxpartSize, MaxMultipartParts and MaxObjectSize are variable too and might be very different in other S3 API implementations.

You might end up with a lot more tests, than you are prepared to handle.

File: s3store/s3store_test.go (hunk 19 above)
Comment (trickkiste) [reply s3store/s3store_test.go]: It works. The output is correct.
But of course, we can pump this up to the latest testify version and switch to assert.Falsef.

File: s3store/s3store_test.go (hunk 18 above)
Comment (trickkiste) [reply s3store/s3store_test.go]: I believe there is a difference between testing something like WriteChunk, which by itself depends on other APIs, functions and methods and a helper function with only one input and one output and no additional level of abstraction.

For WriteChunk the quality of the data on the input can differ quite significantly, both in value, but also in its structural composition. And moreover, we depend on other functions, methods and interfaces below it.

While CalcOptimalPartSize just takes a number and will return another one, or an error. But it will not venture off into other layers of abstraction.

File: s3store/s3store_test.go (hunk 19 above)
Comment (Acconut) [reply s3store/s3store_test.go]: Sorry, my bad. I wasn't aware that assert.False also accepts a formattable string. So everything is fine here :)

**Comment by Acconut** [comment]
Merged in 2df5d116725f952da53d8a9b2a75fab0df784833, thanks :)
